lib/compress/zstd_ldm.c

```diff
@@ -466,71 +468,92 @@ static void ZSTD_ldm_reduceTable(ldmEntry_t* const table, U32 const size,
+        /* 2. We enforce the maximum offset allowed.
+         *
+         * kMaxChunkSize should be small enough that we don't lose too much of
          * the window through early invalidation.
          * TODO: * Test the chunk size.
          *       * Try invalidation after the sequence generation and test the
          *         the offset against maxDist directly.
          */
         ZSTD_window_enforceMaxDist(&ldmState->window, istart + chunkEnd,
                                    maxDist);
-        nbSeq += ZSTD_ldm_generateSequences_internal(
-            ldmState, sequences + nbSeq, params, istart + chunkStart, chunkSize,
-            extDict);
+        /* 3. Generate the sequences for the chunk, and get newLeftoverSize. */
+        newLeftoverSize = ZSTD_ldm_generateSequences_internal(
+            ldmState, sequences, params, istart + chunkStart,
+            chunkSize);
+        if (ZSTD_isError(newLeftoverSize))
+            return newLeftoverSize;
+        /* 4. We add the leftover literals from previous iterations to the first
+         *    newly generated sequence, or add the `newLeftoverSize` if none are
+         *    generated.
+         */
+        /* Prepend the leftover literals from the last call */
+        if (prevPos < sequences->pos) {
+            sequences->seq[prevPos].litLength += (U32)leftoverSize;
+            leftoverSize = newLeftoverSize;
+        } else {
+            assert(newLeftoverSize == chunkSize);
+            leftoverSize += chunkSize;
+        }
     }
-    return nbSeq;
+    /* Return the number of generated sequences */
+    return 0;
 }
 
-#if 0
 /**
  * If the sequence length is longer than remaining then the sequence is split
  * between this block and the next.
  *
  * Returns the current sequence to handle, or if the rest of the block should
  * be literals, it returns a sequence with offset == 0.
  */
-static rawSeq maybeSplitSequence(rawSeq* sequences, size_t* nbSeq,
-                                 size_t const seq, size_t const remaining,
-                                 U32 const minMatch)
+static rawSeq maybeSplitSequence(rawSeqStore_t* rawSeqStore,
+                                 U32 const remaining, U32 const minMatch)
 {
-    rawSeq sequence = sequences[seq];
+    size_t const pos = rawSeqStore->pos;
+    rawSeq sequence = rawSeqStore->seq[rawSeqStore->pos];
     assert(sequence.offset > 0);
     /* Handle partial sequences */
     if (remaining <= sequence.litLength) {
         /* Split the literals that we have out of the sequence.
          * They will become the last literals of this block.
          * The next block starts off with the remaining literals.
          */
-        sequences[seq].litLength -= remaining;
-        *nbSeq = seq;
+        rawSeqStore->seq[pos].litLength -= remaining;
         sequence.offset = 0;
     } else if (remaining < sequence.litLength + sequence.matchLength) {
         /* Split the match up into two sequences. One in this block, and one
```

Comment:
What about splitting at the boundary between literals and match ?

This would create variable size blocks. 
But I suspect the rest of the code is not yet ready for that ?
note : could be left as a refinement for later.

Reply:
Yeah, that would take some refactoring of the block compression code. `ZSTD_compressContinue_internal()`, `ZSTD_compress_frameChunk()`, and `ZSTD_compressBlock()` all expect `ZSTD_compressBlock_internal()` to compress all the input. It would also require a way to request not to do this, since it could be the last block in the frame, or a raw block, so all the input must be consumed.